### ISSUE TITLE
High disk utilisation after network destination reconnect with reliable disk buffer

### ISSUE THREAD

**Issue by L4rS6** [syslog-ng/syslog-ng#3551]
# syslog-ng
## Version of syslog-ng
```
syslog-ng 3 (3.30.1)
Config version: 3.29
Installer-Version: 3.30.1
Revision:
Compile-Date: Nov 22 2020 00:16:57
Module-Directory: /usr/local/lib/syslog-ng
Module-Path: /usr/local/lib/syslog-ng
Include-Path: /usr/local/share/syslog-ng/include
Available-Modules: appmodel,json-plugin,stardate,affile,system-source,disk-buffer,timestamp,http,map-value-pairs,examples,confgen,afstomp,kvformat,afuser,csvparser,afprog,basicfuncs,xml,afsocket,hook-commands,tfgetent,secure-logging,syslogformat,add-contextual-data,dbparser,cef,pseudofile,graphite,linux-kmsg-format,cryptofuncs,tags-parser,azure-auth-header
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off
```

## Platform
`FreeBSD 11.4`

# Issue
After network reconnect which uses a reliable disk buffer syslog-ng utilise the disk nearly to 100%. This problem can be fixed by disabling the disk buffer, restarting syslog-ng and reenabling the disk buffer.

## Steps to reproduce

1. Start logging: `while true; do logger -p kern.crit test; sleep 0.001; done`
2. Check disk utilisation: `gstat` (Disk has nearly no utilisation)
3. Drop TCP connection to logging host: `tcpdrop 10.0.0.1 5555 10.0.0.2 514`
4. Wait until syslog-ng reconnects to remote logging host (default 1 minute)
5. Check disk utilisation: `gstat` (Disk has nearly 100% utilisation)
6. Issue can be fixed by: Remove disk buffer part in configuration --> `service syslog-ng restart` --> Readd disk buffer part in configuration --> `service syslog-ng restart`

## Configuration
```
@version:3.30

options {
  create_dirs(yes);
  keep_hostname(yes);
  threaded(yes);
  log_fifo_size(32000);
  frac_digits (3);
};

source s_local {
  system();
  internal();
 };

destination d_network {
  syslog(
    "logging.myhost"
    port(514)
    disk-buffer(
      disk-buf-size(1G)
      reliable(yes)
    )
  );
};

# Add hostgroup
rewrite r_hostgroup { set("testhostgroup", value(".SDATA.hostgroup@18372.4")); };
log { source(s_local); rewrite(r_hostgroup); destination(d_network); flags(flow-control); };
```

## Input and output logs (if possible)
Input generator: `while true; do logger -p kern.crit test; sleep 0.001; done`

**Comment by alltilla** [comment]
Hi @L4rS6,

There are 2 things to consider:
 1. Writing to and reading from the `disk-buffer` can be slow because of disk IO limitations.
 2. In that 1 minute, the `disk-buffer` can get quite large.

After the destination is reconnected, two things happen simultaneously:
 1. Old logs are read from the `disk-buffer`.
 2. New logs are arriving and getting written to the `disk-buffer`.

In a bad case this cannot resolve itself, because the logs are arriving with a higher rate, than we can read old ones from the `disk-buffer`.
If the inflow stops for enough time, we can read all logs from the `disk-buffer`, and then we can continue the log processing on the "fast" path, without writing to and reading from the `disk-buffer`.

Removing and readding the `disk-buffer` will cause you to lose logs.

I assume you are trying to create a config for a production environment and you are testing how syslog-ng operates. To fine-tune reliability and processing rate, it is necessary to know more of the production environment's parameters, like its incoming EPS (event per second), EPS consistency (does it have big spikes, or it is more like a constant flow), disk IO limitations, network limitations and reliabilities, message processing speed (which is affected by the configuration, like parsing, filtering and rewriting), etc...

In this case I would advise to check whether the `disk-buffer` empties eventually. If not, then `flow-control` and kernel level buffering could take care of the reliability to some degree (until the kernel buffer becomes full).

**Comment by L4rS6** [comment]
Hi @alltilla 

Thank you for your quick reply.

> There are 2 things to consider:
> 
> 1. Writing to and reading from the disk-buffer can be slow because of disk IO limitations.
> 2. In that 1 minute, the disk-buffer can get quite large.

I totally agree and I'm aware that the reliable disk-buffer is very IO intensive and that the disks IO limitation can be the bottleneck.


> After the destination is reconnected, two things happen simultaneously:
> 
> Old logs are read from the disk-buffer.
> New logs are arriving and getting written to the disk-buffer.
> In a bad case this cannot resolve itself, because the logs are arriving with a higher rate, than we can read old ones from the disk-buffer.

I agree, but I just tested it again and even if I stop my log generator until the disk-buffer gets empty:
```
# cat /var/db/syslog-ng-00001.rqf
SLRQ
```

and after that I restarted the log generator and the disk utilisation goes up to nearly 100%. The disk utilisation has dropped to zero as soon as I dropped the TCP connection to the central logging server, but got up as soon as the TCP connection has been reestablished (and so the disk-buffer is used for writing and reading).


> If the inflow stops for enough time, we can read all logs from the disk-buffer, and then we can continue the log processing on the "fast" path, without writing to and reading from the disk-buffer.

Is there really a difference? Because if I set reliable to yes than all logs will first be written to de disk-buffer and only after successful transmission to the central logging server removed from it, or am I wrong?



> I assume you are trying to create a config for a production environment and you are testing how syslog-ng operates. To fine-tune reliability and processing rate, it is necessary to know more of the production environment's parameters, like its incoming EPS (event per second), EPS consistency (does it have big spikes, or it is more like a constant flow), disk IO limitations, network limitations and reliabilities, message processing speed (which is affected by the configuration, like parsing, filtering and rewriting), etc...

Yes and no, we have already our production configuration and are running syslog-ng on many hosts successfully. We just hit that "problem" from time to time (e.g. if the central logging server requires a restart and all TCP connections get dropped).

> In this case I would advise to check whether the disk-buffer empties eventually. If not, then flow-control and kernel level buffering could take care of the reliability to some degree (until the kernel buffer becomes full).

As already mentioned above I just tried that and I can confirm that the disk-buffer gets emptied but the high IO usage is still there.

**Comment by alltilla** [comment]
Yes you are right about the `reliable disk-buffer` part, I did not notice it was set to `yes` in your config, my bad.

Thanks for trying out stopping the log generation until the `disk-buffer` becomes empty.
I tried to replicate the same on my Ubuntu 20.04 machine, but did not observe this behavior. I will create a FreeBSD 11.4 VM, and try it there.

I will keep you updated!

**Comment by alltilla** [comment]
I tried it on a FreeBSD VM, and had a different but still strange observation. I tried it with both the `logger` infinite loop (1000 EPS), and with `loggen`, too (~10.000 EPS).

 1. Generate messages -> disk usage is ~50%
 2. Disconnect the destination -> messages are getting written to the `disk-buffer`, disk usage is ~0%
 3. Reconnect the destination -> messages are getting read from and being written to the `disk-buffer`, but the disk-usage is ~0%
 4. All messages are emptied from the `disk-buffer` -> disk usage is ~50%

In the meantime I checked the queue file with `dqtool`, and my observation was, that I only had high disk usage, when the `disk-queue` was empty, and the messages were written to it and immediately read from it. When I had several messages in the `disk-buffer` and I was still writing new messages to it and reading old messages from it, the disk usage was nearly 0%.

This seems counter-intuitive for me on the first glance, not saying it is buggy, but strange. Anyways, I will investigate the reason behind it.

In the meantime, I think it would be interesting if you could check the `disk-buffer`'s length before the destination disconnects and the disk usage us ~0%, and after the destination reconnects and the disk usage is ~100%. The command is `dqtool info /var/db/syslog-ng-00000.rqf`

If you want to test with a slightly higher load, you can use `loggen`:
 1. Add `udp(port(12341));` to the `s_local` source.
 2. Execute command: `loggen --rate=1000 --active-connections=10 --dgram --permanent localhost 12341`

**Comment by bazsi** [comment]
Don't forget the page cache, you might have GBs of the disk buffer in
memory.

**Comment by L4rS6** [comment]
I just tested it with the loggen utility:
```
# loggen --rate=1000 --active-connections=10 --dgram --permanent localhost 12341
...
count=18501, rate = 9171.53 msg/sec
count=23071, rate = 9131.29 msg/sec
...
```

During that the disk utilisation was between 0-50% and the disk-buffer size before the disconnect was flapping between 0 some messages:
```
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='0', size='0'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='5', size='2990'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='3', size='1794'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='1', size='598'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='0', size='0'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='1', size='598'
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='0', size='0'
```


after that I dropped the TCP session and de disk-buffer grew:
```
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='59896', size='35817456'
```


and after successful reconnect all messages have been delivered and the disk-buffer has been emptied:
```
# dqtool info /var/db/syslog-ng-00001.rqf
Reliable disk-buffer state loaded; filename='/var/db/syslog-ng-00001.rqf', queue_length='0', size='0'
```

And after that point the disk utilisation went up to nearly 100%.

**Comment by L4rS6** [comment]
@alltilla Have you been able to find something which is causing those strange high IO utilisation?

**Comment by alltilla** [comment]
@L4rS6 Sorry, I did not get to work on this in the previous week. I will check it as soon as I can.

**Comment by alltilla** [comment]
Hi @L4rS6,

Sorry for the long delay!

I did some profiling with `truss`, although my disk usage was never 100%, I observed, that we were staying in `ftruncate()` for a while. Can you give the `truss` output of your run?

The reproduction steps are the same, but start `syslog-ng` with `truss -D -S -c syslog-ng -F`.

Thanks!
Cheers,
Attila

**Comment by L4rS6** [comment]
Hi @alltilla 

Thank you for your investigation.

I started syslog-ng with your command above and tried to reproduced the problem as [here](https://github.com/syslog-ng/syslog-ng/issues/3551#issuecomment-769681471)

```
# truss -D -S -c syslog-ng -F
syscall                     seconds   calls  errors
thr_new                 0.000899767      11       0
kqueue                  0.000335092      10       0
geteuid                 0.000011851       1       0
getcontext              0.000013375       1       0
getpid                  0.000303044      31       0
__sysctl                0.000520575      28       0
issetugid               0.000165923      13       0
write                   0.000192104      16       0
unlink                  0.000249335       5       1
thr_set_name            0.000033630       2       0
thr_self                0.000428931       9       0
sysarch                 0.000026013       2       0
stat                    0.001636941      74       4
socket                  0.000228043      14       0
sigreturn               0.000083568      13       0
sigprocmask             0.003901572     323       0
sigaction               0.000140449      10       0
shutdown                0.000050295       2       1
setsockopt              0.000387765      35      10
setrlimit               0.000026776       2       0
sendto                 12.158888051 1446654       1
rtprio_thread           0.000012857       1       0
rename                  0.000091589       1       0
recvmsg                 6.870624310  738638   15318
recvfrom                0.000025899       2       0
readlink                0.000058027       1       1
read                    0.000881021      43       1
pwrite                 18.859552905 1446653       0
pread                  18.086744683 2169979       0
poll                    0.000170189       2       0
pipe2                   0.000070819       5       0
openat                  0.002625982     131      58
open                    0.000665099      30       1
nanosleep               0.000127215       1       0
munmap                  0.002660226     138       0
mprotect                0.000167800      11       0
mmap                    0.005659286     325       0
madvise                 0.000011486       1       0
lseek                   0.000014601       1       0
listen                  0.000013619       1       0
kevent                310.137413875  355578       0
ioctl                   0.000048678       4       4
getsockopt              0.000108418       9       0
getsockname             0.000141368      12       0
getrlimit               0.000039167       3       0
getdirentries           0.000076203       2       0
ftruncate               0.004603133       5       0
fstatfs                 0.000014392       1       0
fstat                   0.001588905     104       0
fcntl                   0.001818788     115       0
connect                 0.000197029      10       2
close                   0.002287840     127       0
chmod                   0.000056188       2       0
bind                    0.000152166       6       0
access                  0.000169575       7       0
accept                  0.000037570       2       0
_umtx_op               69.217432778  207835       0
                      ------------- ------- -------
                      435.364886786 6367042   15402
```

What I currently really don't understand is that the disk utilisation is zero when I start syslog-ng with the truss command above and try to reproduce the high disk utilisation. 

Also very strange is that the disk utilisation drops to nearly zero as soon as I attach to the syslog-ng process (command: `truss -D -S -c -p 39955`) when the disk utilisation before was ~100%. What do you think about that?

Thank you in advance.





Just let me know if you need further information.

**Comment by alltilla** [comment]
Hi @L4rS6,
Sorry, I was busy with other tasks. I will check this, next week.

**Comment by alltilla** [comment]
Hi,

Thank you for your patience!

So `truss` definitely slows down the `syslog-ng` process enough, to operate below the disk IO limit.

Your and my `truss` output differ greatly in regards of the `ftruncate` calls, mine had more `ftruncate` than `recvmsg`, you only had 5. We have a hunch, that maybe this causes the problem. When the destination disconnects, a huge load of messages are being written to the disk-buffer. After the destination reconnects, and all messages in the disk-buffer are processed, we should truncate the file, and start to write the new messages to the beginning. It seems like, that this truncation does not happen for you, so every message is written and getting read with a big offset in the file. The only thing we do not understand is why would this cause a problem, as we are using `pread` and `pwrite`, which should easily jump to the specified offset in the file.

To validate this hypothesis, I would ask you to gather some more info:
 * Run `truss` without `-c`, so we can see how each syscall was executed.
 * When you have problematic disk-buffer (you have 100% disk load), stop `syslog-ng` and send us the disk-buffer file (`/var/db/syslog-ng-00001.rqf`). Note that this file contains possibly classified log messages, so only send it from the test run, and not the one which is used in the production.
 * We would like to see a debug run of `syslog-ng`. For this you need to start `syslog-ng` with `syslog-ng -Fedtv`, which starts it in the foreground with the most verbose internal logging. Before this, please stop the `syslog-ng` running as a service.
 * For one run, please use `iostat` instead of `gstat` to monitor the disk usage.

Also can you tell us more about your file system? Is it hardware or virutal? Is it local, or NFS?

Thank you!
Attila

**Comment by L4rS6** [comment]
Hi Attila

Sorry for the long delay, but finally I was able to collecte all the requested information.

I attached all the requested outputs:

- truss without `-c` [truss_output.txt.zip](https://github.com/syslog-ng/syslog-ng/files/6271311/truss_output.txt.zip)
- disk-buffer file [syslog-ng-00002.rqf.zip](https://github.com/syslog-ng/syslog-ng/files/6271318/syslog-ng-00002.rqf.zip)
- output of `syslog-ng -Fedtv` [syslog_ng_debug.txt](https://github.com/syslog-ng/syslog-ng/files/6271319/syslog_ng_debug.txt)
- output of gstat and iostat at the same time: [iostat.txt](https://github.com/syslog-ng/syslog-ng/files/6271320/iostat.txt) [gstat.txt](https://github.com/syslog-ng/syslog-ng/files/6271321/gstat.txt)

Yes sure these are the specs:

- FreeBSD 11.4
- HP ProLiant BL460c Gen9
- Local disks (non SSD)
- Filesystem is a ZFS zpool

Let me know if you need further information.

Thank you in advance!
Lars

**Comment by L4rS6** [comment]
Hi @alltilla, were you able to find more details with the information and logs above?

**Comment by alltilla** [comment]
Hi @L4rS6,

Yes I have found something.

I investigated `ftruncate()` a bit closer.

Disk-buffer has a read (reading from disk-buffer and sending to destination) and write (receiving from source and writing to disk-buffer) cursor, where the read cursor tries to catch-up to the write cursor. When it reaches it, we know that we processed every message written to the disk-buffer. At that point, we call `ftruncate()` to remove every message from the disk-buffer, and move the cursors to the start.
 1. When syslog-ng does not receive messages with a high frequency, the read cursor always reaches the write cursor, so we call `ftruncate()` for every message. So the number and the frequency `ftruncate()` calls will be equal to the received messages'.
 2. We can increase the message frequency to a point, where the read cursor barely reaches the write cursor for every message. In this case we have the most `ftruncate()` calls per second.
 3. If we incrase the message frequency even more, the read cursor lags behind the write cursor, and we practically never call `ftruncate()`.

So there is a "sweet spot" at point 2, where we have the most `ftruncate()` calls per second.
I have found a bug, where we called `ftruncate()` twice as much as it was necessary. This, combined with the "sweet spot" might cause an unnecessarily high disk usage.

We observed something similar at your measurements, when you turned on `truss`, you have had a really small number of `ftruncate()` calls, and your IO usage was small. I presume, that when `truss` is not started, the read cursor can catch up to the write cursor and we call `ftruncate()` frequently, so we have a high disk usage. I cannot say for certain, that `ftruncate()` and the high disk usage have a causal relation, but they correlate.

I would like to provide you a test fix. I prepared a patch for FreeBSD ports (`patch-ftruncate`), where I get rid of the unnecessary `ftruncate()` call. You can place it in `/usr/ports/sysutils/syslog-ng-330/files`. There is more info about building syslog-ng from source here: https://www.syslog-ng.com/community/b/blog/posts/syslog-ng-in-freebsd-ports

Can you give it a go, and see if it helps anything?

Thanks!

Cheers,
Attila

`patch-ftruncate`:
```
diff --git modules/diskq/qdisk.c modules/diskq/qdisk.c
index cebe98aa3..d07233690 100644
--- modules/diskq/qdisk.c
+++ modules/diskq/qdisk.c
@@ -201,35 +201,32 @@ qdisk_is_space_avail(QDisk *self, gint at_least)
 }
 
 static gboolean
-_truncate_file(QDisk *self, off_t new_size)
+_truncate_file(QDisk *self, off_t expected_size)
 {
-  gboolean success = TRUE;
-
-  if (ftruncate(self->fd, new_size) < 0)
+  if (ftruncate(self->fd, expected_size) == 0)
     {
-      success = FALSE;
-      off_t file_size = -1;
-
-      struct stat st;
-      if (fstat(self->fd, &st) < 0)
-        {
-          msg_error("truncate file: cannot stat",
-                    evt_tag_error("error"));
-        }
-      else
-        {
-          file_size = st.st_size;
-        }
-
-      msg_error("Error truncating disk-queue file",
-                evt_tag_error("error"),
-                evt_tag_str("filename", self->filename),
-                evt_tag_long("expected-size", new_size),
-                evt_tag_long("file_size", file_size),
-                evt_tag_int("fd", self->fd));
+      self->file_size = expected_size;
+      return TRUE;
     }
 
-  return success;
+  struct stat st;
+  if (fstat(self->fd, &st) < 0)
+    {
+      msg_error("truncate file: cannot stat", evt_tag_error("error"));
+    }
+  else
+    {
+      self->file_size = st.st_size;
+    }
+
+  msg_error("Error truncating disk-queue file",
+            evt_tag_error("error"),
+            evt_tag_str("filename", self->filename),
+            evt_tag_long("expected-size", expected_size),
+            evt_tag_long("file_size", self->file_size),
+            evt_tag_int("fd", self->fd));
+
+  return FALSE;
 }
 
 static gint64
@@ -348,7 +345,10 @@ qdisk_push_tail(QDisk *self, GString *record)
                     evt_tag_long("new size",  self->hdr->write_head));
           _truncate_file(self, self->hdr->write_head);
         }
-      self->file_size = self->hdr->write_head;
+      else
+        {
+          self->file_size = self->hdr->write_head;
+        }
 
       if (_is_qdisk_overwritten(self) && self->hdr->backlog_head  != QDISK_RESERVED_SPACE)
         {
```

**Comment by L4rS6** [comment]
Hi @alltilla 

Thank you for your investigation.

Unfortunately there is no difference with the applied patch. Shall I collect some information?

Thank you in advance.
Lars

**Comment by alltilla** [comment]
Hi @L4rS6 

Can you try it first with this additionally to `patch-ftruncate`, and see if the disk usage is better?
`patch-ftruncate-2`:
```
diff --git modules/diskq/qdisk.c modules/diskq/qdisk.c
index d07233690..7145f7349 100644
--- modules/diskq/qdisk.c
+++ modules/diskq/qdisk.c
@@ -1021,7 +1021,7 @@ qdisk_skip_record(QDisk *self, guint64 position)
 void
 qdisk_reset_file_if_possible(QDisk *self)
 {
-  if (qdisk_is_file_empty(self))
+  if (qdisk_is_file_empty(self) && self->file_size != QDISK_RESERVED_SPACE)
     {
       self->hdr->read_head = QDISK_RESERVED_SPACE;
       self->hdr->write_head = QDISK_RESERVED_SPACE;

```

And if not, please add this too, and start syslog-ng in foreground: `syslog-ng -Fe`, and send me the logs:
`patch-debug-logs`:
```
diff --git modules/diskq/qdisk.c modules/diskq/qdisk.c
index 7145f7349..a3e9760cf 100644
--- modules/diskq/qdisk.c
+++ modules/diskq/qdisk.c
@@ -82,6 +82,7 @@ struct _QDisk
 static gboolean
 pwrite_strict(gint fd, const void *buf, size_t count, off_t offset)
 {
+  msg_error("ALLTILLA: pwrite", evt_tag_int("size", count), evt_tag_int("position", offset));
   ssize_t written = pwrite(fd, buf, count, offset);
   gboolean result = TRUE;
   if (written != count)
@@ -203,6 +204,8 @@ qdisk_is_space_avail(QDisk *self, gint at_least)
 static gboolean
 _truncate_file(QDisk *self, off_t expected_size)
 {
+  msg_error("ALLTILLA: ftruncate", evt_tag_int("file_size", self->file_size), evt_tag_int("expected_size", expected_size));
+
   if (ftruncate(self->fd, expected_size) == 0)
     {
       self->file_size = expected_size;
@@ -378,12 +381,15 @@ qdisk_pop_head(QDisk *self, GString *record)
     {
       guint32 record_length;
       gssize res;
+      msg_error("ALLTILLA: pread (pop_head record_length)", evt_tag_int("size", sizeof(record_length)), evt_tag_int("read_head", self->hdr->read_head));
+
       res = pread(self->fd, (gchar *) &record_length, sizeof(record_length), self->hdr->read_head);
 
       if (res == 0)
         {
           /* hmm, we are either at EOF or at hdr->qout_ofs, we need to wrap */
           self->hdr->read_head = QDISK_RESERVED_SPACE;
+          msg_error("ALLTILLA: pread (pop_head res==0)", evt_tag_int("size", sizeof(record_length)), evt_tag_int("read_head", self->hdr->read_head));
           res = pread(self->fd, (gchar *) &record_length, sizeof(record_length), self->hdr->read_head);
         }
       if (res != sizeof(record_length))
@@ -414,6 +420,7 @@ qdisk_pop_head(QDisk *self, GString *record)
         }
 
       g_string_set_size(record, record_length);
+      msg_error("ALLTILLA: pread (pop_head record)", evt_tag_int("size", record_length), evt_tag_int("read_head", self->hdr->read_head));
       res = pread(self->fd, record->str, record_length, self->hdr->read_head + sizeof(record_length));
       if (res != record_length)
         {
@@ -656,6 +663,7 @@ string_reached_memory_limit(GString *string)
 static gboolean
 qdisk_write_serialized_string_to_file(QDisk *self, GString const *serialized, gint64 *offset)
 {
+  msg_error("ALLTILLA: lseek");
   *offset = lseek(self->fd, 0, SEEK_END);
   if (!pwrite_strict(self->fd, serialized->str, serialized->len, *offset))
     {
@@ -993,6 +1001,8 @@ gssize
 qdisk_read(QDisk *self, gpointer buffer, gsize bytes_to_read, gint64 position)
 {
   gssize res;
+  msg_error("ALLTILLA: pread (qdisk_read)", evt_tag_int("size", bytes_to_read), evt_tag_int("position", position));
+
   res = pread(self->fd, buffer, bytes_to_read, position);
   if (res <= 0)
     {

```

Thanks!

**Comment by L4rS6** [comment]
Hi @alltilla 

I added the 2nd patch and it didn't make any difference.

After that I added your 3rd patch and let syslog-ng running in foreground. Interestingly the disk utilisation didn't go up to ~95% after the TCP connection has been reestablished (might the logging had slowed it down?). That's why I restarted syslog-ng in foreground and then the disk utilisation went up to ~95% (see 2nd logfile syslog-ng2.log.zip).

[syslog-ng.log.zip](https://github.com/syslog-ng/syslog-ng/files/6440471/syslog-ng.log.zip)

[syslog-ng2.log.zip](https://github.com/syslog-ng/syslog-ng/files/6440505/syslog-ng2.log.zip)

Thank you!

**Comment by alltilla** [comment]
Hi @L4rS6,

Thank you for the logs, they are really helpful.

It can be seen, that in case of the high disk usage logs (`syslog-ng2.log`), there were some really big delays between the syscalls.
For example:
```
[2021-05-07T11:21:15.621562] ALLTILLA: ftruncate; file_size='13664', expected_size='4096'
[2021-05-07T11:21:37.179846] ALLTILLA: pwrite; size='4', position='4096'
```

For comparison I will attach some graphs from the `ftruncate()`, `pwrite()` and `pread()` logs in `syslog-ng2.log` and in `syslog-ng.log` (no high disk usage).

These logs can indicate, that the `ftruncate()` call took a lot of time, or that `ftruncate()` finished quickly, but some other thing caused following the `pwrite()` call to be delayed.
I would like to give you 2 patches, one which logs before and immediately after the `ftruncate()` call, and one that completely removes `ftruncate()`.

Please build and run your case with each patch, only applying 1 patch at a time (the previous patches are not needed now) and send me the logs.

## Patches:
[disable-and-before-after-logs.zip](https://github.com/syslog-ng/syslog-ng/files/6452715/disable-and-before-after-logs.zip)

## Graphs:
Y-axis is in seconds.
### syslog-ng2.log
#### ftruncate
![1-ftruncate](https://user-images.githubusercontent.com/45236571/117663608-24316500-b1a1-11eb-9bfe-fa6024b02591.png)
#### pread
![1-pread](https://user-images.githubusercontent.com/45236571/117663613-272c5580-b1a1-11eb-803e-285bb16b7c0c.png)
#### pwrite
![1-pwrite](https://user-images.githubusercontent.com/45236571/117663623-28f61900-b1a1-11eb-88c1-52b4b926884b.png)
### syslog-ng.log
#### ftruncate
![2-ftruncate](https://user-images.githubusercontent.com/45236571/117663628-2b587300-b1a1-11eb-9490-d678d45fdfc0.png)
#### pread
![2-pread](https://user-images.githubusercontent.com/45236571/117663637-2d223680-b1a1-11eb-81f0-6f246f02b80a.png)
#### pwrite
![2-pwrite](https://user-images.githubusercontent.com/45236571/117663644-2eebfa00-b1a1-11eb-8cb6-77c12930cc98.png)

We can see some cases, where `pread()` and `pwrite()` "took" 4-5 seconds, those might contribute to the disk usage, but the same is true for these measurements as well, they are just the time between the start of 2 syscalls, so they are not the time we spent in them.

Also, we can see, that the `ftruncate()` calls stopped for 4 minutes at `2021-05-07T11:12:24.176844` for the not high disk usage case, while in the high disk usage case `ftruncate()` was called regularly. This makes me assume, that there is something behind `ftruncate()`, not sure, but we will know more with the new logs.

Thank you again, for your extensive help!

Cheers,
Attila

**Comment by L4rS6** [comment]
Hi @alltilla 

I added one after the other of your patches. 

With the disable-ftruncate patch there was nearly no disk utilisation and with the ftruncate-before-after-logs patch the disk utilisation went up to ~95%. For further details see the attached logs. I also restarted syslog-ng with both patches to see if there is any difference, but there was the same behaviour as before the restart.

[syslog-ng-disable-ftruncate.log.zip](https://github.com/syslog-ng/syslog-ng/files/6457979/syslog-ng-disable-ftruncate.log.zip)

[syslog-ng-ftruncate-before-after-logs.log.zip](https://github.com/syslog-ng/syslog-ng/files/6458020/syslog-ng-ftruncate-before-after-logs.log.zip)

Thank you for the detailed information and graphs!

Cheers
Lars

**Comment by alltilla** [comment]
Hi @L4rS6,

Thanks for the logs!

I am glad to see, that disabling `ftruncate()` resolves the issue. So it was caused by that after all.
See this graph, where the X axis is the time elapsed since the first `ftruncate()` after reconnecting to the destination, and the Y axis is the percentage of time spent in `ftruncate()` since that:

![Screenshot from 2021-05-11 15-19-43](https://user-images.githubusercontent.com/45236571/117821952-5e1b6d80-b26c-11eb-8c00-1fe6f131c8e2.png)

It can be clearly seen, that after the reconnect's first `ftruncate()`, we spend nearly 100% of the time in `ftruncate()`.
The only difference between the start and the time after the reconnect, is that after reconnecting, the file becomes quite large, but the first `ftruncate()` should and does take care of that. But for some reason concurrent `ftruncate()` calls still take much time. And as you said, stopping and starting `syslog-ng` does not fix this problem, it only gets fixed, if you completely remove the disk-buffer from the config and add it back again.

I don't know what we could do with the slow `ftruncate()` calls on `syslog-ng` side, other than not calling it with this high frequency (possibly per message), but only during a reload. This would cause the disk-buffer files to be possibly maximal size, even if they could be way smaller, but we would save a lot of syscalls. I can imagine a config option, which would change our truncating pattern from "optimize-file-size" to "optimize-performance", or something like that.

[patch-ftruncate-optimize-perf.zip](https://github.com/syslog-ng/syslog-ng/files/6459436/patch-ftruncate-optimize-perf.zip)

I will speak with the team about this problem, maybe they have some idea, of why does `ftruncate()` take this much time. In the meantime, I send you a patch, where `ftruncate()` will only be called less frequently. I would like to ask you, to try it, send messages, disconnect and reconnect the destination, do some `syslog-ng` reloads, and restarts, and see how it behaves. I expect low disk usage, and higher disk-buffer size, which get small sometimes runtime, or when reloading or stopping `syslog-ng`. Note, that this is a very quick PoC code, you should not use it in production.

If I managed to talk with the team, I will reach out to you!

Cheers,
Attila

**Comment by alltilla** [comment]
Hi @L4rS6,

To isolate the problem, I wrote a test code, which does the following:
 1. Creates a file with 4096 bytes padding.
 2. Writes a 1024 bytes sized data to a file.
 3. Reads back that 1024 bytes.
 4. Truncates the file to the 4096 bytes padding.
 5. Does steps 2-4 10*100.000 times, logging each of the 10 iteration's time spent, and time spent in `ftruncate()`.
 6. Writes 1 GB of data to the file.
 7. Reads back that 1 GB of data.
 8. Truncates the file to the 4096 bytes padding.
 9. Repeats step 5.

You can modify the parameters in the beginning of the source code.
With this I am planning to prove, that once the file becomes large, the following `ftruncate()` calls take a lot of time.

Example run at my computer:
```
$ gcc ftruncate-test.c -o ftruncate-test && ./ftruncate-test /tmp/disk-buffer-file
Writing, reading 100000 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 0.830026 seconds, from which ftruncate() took 0.486980 seconds.
  [2/10] DONE in 0.842110 seconds, from which ftruncate() took 0.497163 seconds.
  [3/10] DONE in 0.854882 seconds, from which ftruncate() took 0.500999 seconds.
  [4/10] DONE in 0.825156 seconds, from which ftruncate() took 0.485670 seconds.
  [5/10] DONE in 0.830391 seconds, from which ftruncate() took 0.489473 seconds.
  [6/10] DONE in 0.843066 seconds, from which ftruncate() took 0.495798 seconds.
  [7/10] DONE in 0.918503 seconds, from which ftruncate() took 0.536093 seconds.
  [8/10] DONE in 0.853550 seconds, from which ftruncate() took 0.500094 seconds.
  [9/10] DONE in 0.835046 seconds, from which ftruncate() took 0.493158 seconds.
  [10/10] DONE in 0.841506 seconds, from which ftruncate() took 0.493619 seconds.
Writing 1048576 * 1024 bytes, then reading it, then truncating to 4096.
  DONE in 2.248595 seconds.
Writing, reading 100000 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 0.820598 seconds, from which ftruncate() took 0.482383 seconds.
  [2/10] DONE in 0.833353 seconds, from which ftruncate() took 0.490249 seconds.
  [3/10] DONE in 0.834549 seconds, from which ftruncate() took 0.490391 seconds.
  [4/10] DONE in 0.845641 seconds, from which ftruncate() took 0.497339 seconds.
  [5/10] DONE in 0.838015 seconds, from which ftruncate() took 0.494708 seconds.
  [6/10] DONE in 0.870128 seconds, from which ftruncate() took 0.509075 seconds.
  [7/10] DONE in 0.845684 seconds, from which ftruncate() took 0.494878 seconds.
  [8/10] DONE in 0.843265 seconds, from which ftruncate() took 0.494923 seconds.
  [9/10] DONE in 0.829744 seconds, from which ftruncate() took 0.487091 seconds.
  [10/10] DONE in 0.846455 seconds, from which ftruncate() took 0.498038 seconds.
```
You can see, that in my case the 2nd round of per message write, read, truncate took the same amount of time, that in the 1st round, so the big file did not affect it, but I hope to see a significant slow on your computer between the 1st and 2nd round.

Can you please try it yourself? Thanks!

[ftruncate-test.zip](https://github.com/syslog-ng/syslog-ng/files/6465734/ftruncate-test.zip)

Cheers,
Attila

**Comment by alltilla** [comment]
Thanks to @MrAnno, I managed to test this code on a FreeBSD 11.4 with ZFS, and got the following:
```
Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 0.003932060 seconds, from which ftruncate() took 0.001965817 seconds.
  [2/10] DONE in 0.003592179 seconds, from which ftruncate() took 0.001781190 seconds.
  [3/10] DONE in 0.003914860 seconds, from which ftruncate() took 0.001928066 seconds.
  [4/10] DONE in 0.003797470 seconds, from which ftruncate() took 0.001901012 seconds.
  [5/10] DONE in 0.003952670 seconds, from which ftruncate() took 0.001904070 seconds.
  [6/10] DONE in 0.003954710 seconds, from which ftruncate() took 0.001944983 seconds.
  [7/10] DONE in 0.003835820 seconds, from which ftruncate() took 0.001955091 seconds.
  [8/10] DONE in 0.003925920 seconds, from which ftruncate() took 0.001883405 seconds.
  [9/10] DONE in 0.004426200 seconds, from which ftruncate() took 0.002025593 seconds.
  [10/10] DONE in 0.004076190 seconds, from which ftruncate() took 0.001991997 seconds.
Writing 1048576 * 1024 bytes, then reading it, then truncating to 4096.
  DONE in 18.376288550 seconds.
Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 2.794785770 seconds, from which ftruncate() took 2.791617650 seconds.
  [2/10] DONE in 2.731177320 seconds, from which ftruncate() took 2.727718070 seconds.
  [3/10] DONE in 2.928945670 seconds, from which ftruncate() took 2.925396520 seconds.
  [4/10] DONE in 2.951249930 seconds, from which ftruncate() took 2.947725080 seconds.
  [5/10] DONE in 2.693664760 seconds, from which ftruncate() took 2.690457570 seconds.
  [6/10] DONE in 2.644540560 seconds, from which ftruncate() took 2.641382890 seconds.
  [7/10] DONE in 2.791897620 seconds, from which ftruncate() took 2.788706690 seconds.
  [8/10] DONE in 2.547320000 seconds, from which ftruncate() took 2.544186910 seconds.
  [9/10] DONE in 2.069752740 seconds, from which ftruncate() took 2.066673740 seconds.
  [10/10] DONE in 2.602135020 seconds, from which ftruncate() took 2.598901120 seconds.
```

I can see a ~1000 multiplier between the 2 rounds. If you observe the same, I think we can conclude, that it is not a `syslog-ng` fault.

Nevertheless, it can be seen, that on my computer, `ftruncate()` still took more than half of the time of the `pwrite()` -> `pread()` -> `ftruncate()` iteration. Calling it per message has a high cost, and should be changed. If we manage to limit it to `syslog-ng` stop or reload, everyone can profit from it, FreeBSD with ZFS the most.

**Comment by L4rS6** [comment]
Hi @alltilla 

First of all thank you for your explanations.

Finally I was able to test your patch and I still can see that the disk utilisation goes up to at about 95%. Might you have forgotten something in your patch?

Looking forward to hear from you again what's the teams opinion.

Cheers,
Lars




> Hi @L4rS6,
> 
> Thanks for the logs!
> 
> I am glad to see, that disabling `ftruncate()` resolves the issue. So it was caused by that after all.
> See this graph, where the X axis is the time elapsed since the first `ftruncate()` after reconnecting to the destination, and the Y axis is the percentage of time spent in `ftruncate()` since that:
> 
> ![Screenshot from 2021-05-11 15-19-43](https://user-images.githubusercontent.com/45236571/117821952-5e1b6d80-b26c-11eb-8c00-1fe6f131c8e2.png)
> 
> It can be clearly seen, that after the reconnect's first `ftruncate()`, we spend nearly 100% of the time in `ftruncate()`.
> The only difference between the start and the time after the reconnect, is that after reconnecting, the file becomes quite large, but the first `ftruncate()` should and does take care of that. But for some reason concurrent `ftruncate()` calls still take much time. And as you said, stopping and starting `syslog-ng` does not fix this problem, it only gets fixed, if you completely remove the disk-buffer from the config and add it back again.
> 
> I don't know what we could do with the slow `ftruncate()` calls on `syslog-ng` side, other than not calling it with this high frequency (possibly per message), but only during a reload. This would cause the disk-buffer files to be possibly maximal size, even if they could be way smaller, but we would save a lot of syscalls. I can imagine a config option, which would change our truncating pattern from "optimize-file-size" to "optimize-performance", or something like that.
> 
> [patch-ftruncate-optimize-perf.zip](https://github.com/syslog-ng/syslog-ng/files/6459436/patch-ftruncate-optimize-perf.zip)
> 
> I will speak with the team about this problem, maybe they have some idea, of why does `ftruncate()` take this much time. In the meantime, I send you a patch, where `ftruncate()` will only be called less frequently. I would like to ask you, to try it, send messages, disconnect and reconnect the destination, do some `syslog-ng` reloads, and restarts, and see how it behaves. I expect low disk usage, and higher disk-buffer size, which get small sometimes runtime, or when reloading or stopping `syslog-ng`. Note, that this is a very quick PoC code, you should not use it in production.
> 
> If I managed to talk with the team, I will reach out to you!
> 
> Cheers,
> Attila

**Comment by L4rS6** [comment]
Hi @alltilla 

I just let the script run and there is a very big difference between the two runs:

```
# ./a.out /var/db/disk-buffer-file
Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 0.001419848 seconds, from which ftruncate() took 0.000757713 seconds.
  [2/10] DONE in 0.001406707 seconds, from which ftruncate() took 0.000760646 seconds.
  [3/10] DONE in 0.001396399 seconds, from which ftruncate() took 0.000756637 seconds.
  [4/10] DONE in 0.001397363 seconds, from which ftruncate() took 0.000754597 seconds.
  [5/10] DONE in 0.001406701 seconds, from which ftruncate() took 0.000760628 seconds.
  [6/10] DONE in 0.001402528 seconds, from which ftruncate() took 0.000756179 seconds.
  [7/10] DONE in 0.001396875 seconds, from which ftruncate() took 0.000752724 seconds.
  [8/10] DONE in 0.001398534 seconds, from which ftruncate() took 0.000756915 seconds.
  [9/10] DONE in 0.001414452 seconds, from which ftruncate() took 0.000773939 seconds.
  [10/10] DONE in 0.001388983 seconds, from which ftruncate() took 0.000751167 seconds.
Writing 1048576 * 1024 bytes, then reading it, then truncating to 4096.
  DONE in 7.544811855 seconds.
Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
  [1/10] DONE in 45.622374862 seconds, from which ftruncate() took 45.620827242 seconds.
  [2/10] DONE in 38.919319532 seconds, from which ftruncate() took 38.917969258 seconds.
  [3/10] DONE in 37.541376260 seconds, from which ftruncate() took 37.540045645 seconds.
  [4/10] DONE in 42.530202600 seconds, from which ftruncate() took 42.528870524 seconds.
  [5/10] DONE in 42.511952377 seconds, from which ftruncate() took 42.510631410 seconds.
  [6/10] DONE in 37.115356028 seconds, from which ftruncate() took 37.113907170 seconds.
  [7/10] DONE in 45.987602884 seconds, from which ftruncate() took 45.986206752 seconds.
  [8/10] DONE in 43.330486314 seconds, from which ftruncate() took 43.329151288 seconds.
  [9/10] DONE in 39.013205314 seconds, from which ftruncate() took 39.011932633 seconds.
  [10/10] DONE in 40.233253147 seconds, from which ftruncate() took 40.231797749 seconds.
```


Cheers,
Lars

> Hi @L4rS6,
> 
> **EDIT: fixed the test code with proper time measurement**
> 
> To isolate the problem, I wrote a test code, which does the following:
> 
> 1. Creates a file with 4096 bytes padding.
> 2. Writes a 1024 bytes sized data to a file.
> 3. Reads back that 1024 bytes.
> 4. Truncates the file to the 4096 bytes padding.
> 5. Does steps 2-4 10*100 times, logging each of the 10 iteration's time spent, and time spent in `ftruncate()`.
> 6. Writes 1 GB of data to the file.
> 7. Reads back that 1 GB of data.
> 8. Truncates the file to the 4096 bytes padding.
> 9. Repeats step 5.
> 
> You can modify the parameters in the beginning of the source code.
> With this I am planning to prove, that once the file becomes large, the following `ftruncate()` calls take a lot of time.
> 
> Example run at my computer:
> 
> ```
> $ gcc ftruncate-test.c -o ftruncate-test && ./ftruncate-test /tmp/disk-buffer-file
> Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
>   [1/10] DONE in 0.002223723 seconds, from which ftruncate() took 0.001009960 seconds.
>   [2/10] DONE in 0.000978747 seconds, from which ftruncate() took 0.000591587 seconds.
>   [3/10] DONE in 0.000711216 seconds, from which ftruncate() took 0.000425085 seconds.
>   [4/10] DONE in 0.000706371 seconds, from which ftruncate() took 0.000423923 seconds.
>   [5/10] DONE in 0.000704150 seconds, from which ftruncate() took 0.000422382 seconds.
>   [6/10] DONE in 0.000704137 seconds, from which ftruncate() took 0.000423030 seconds.
>   [7/10] DONE in 0.000767069 seconds, from which ftruncate() took 0.000429982 seconds.
>   [8/10] DONE in 0.000705511 seconds, from which ftruncate() took 0.000423481 seconds.
>   [9/10] DONE in 0.000704437 seconds, from which ftruncate() took 0.000422908 seconds.
>   [10/10] DONE in 0.000701327 seconds, from which ftruncate() took 0.000421929 seconds.
> Writing 1048576 * 1024 bytes, then reading it, then truncating to 4096.
>   DONE in 2.464253817 seconds.
> Writing, reading 100 * 1024 bytes messages, truncating the file per message to 4096 bytes.
>   [1/10] DONE in 0.000742420 seconds, from which ftruncate() took 0.000432372 seconds.
>   [2/10] DONE in 0.000708448 seconds, from which ftruncate() took 0.000420383 seconds.
>   [3/10] DONE in 0.000704771 seconds, from which ftruncate() took 0.000419195 seconds.
>   [4/10] DONE in 0.000701254 seconds, from which ftruncate() took 0.000419187 seconds.
>   [5/10] DONE in 0.000703119 seconds, from which ftruncate() took 0.000419106 seconds.
>   [6/10] DONE in 0.000704854 seconds, from which ftruncate() took 0.000419391 seconds.
>   [7/10] DONE in 0.000809425 seconds, from which ftruncate() took 0.000476072 seconds.
>   [8/10] DONE in 0.000940068 seconds, from which ftruncate() took 0.000549031 seconds.
>   [9/10] DONE in 0.000925611 seconds, from which ftruncate() took 0.000552481 seconds.
>   [10/10] DONE in 0.000818691 seconds, from which ftruncate() took 0.000482185 seconds.
> ```
> 
> You can see, that in my case the 2nd round of per message write, read, truncate took the same amount of time, that in the 1st round, so the big file did not affect it, but I hope to see a significant slow on your computer between the 1st and 2nd round.
> 
> Can you please try it yourself? Thanks!
> 
> [ftruncate-test.zip](https://github.com/syslog-ng/syslog-ng/files/6466448/ftruncate-test.zip)
> 
> Cheers,
> Attila

**Comment by alltilla** [comment]
Hi @L4rS6 ,

We talked about this with the team, and we agree with bazsi, that we should not truncate at only reload, I am currently working on a better solution.

45 seconds is a huge number. If you have multiple disk-buffers, even with the "optimized" truncate logic, it will hang syslog-ng for minutes.

I think the solution for your problem is to change the ZFS configuration for the disk-buffer files. I am not really familiar with ZFS, but I have seen multiple guides of how to store big files or databases on ZFS, I suggest to investigate that. You can use the test code to validate your settings, and when the test code behaves as you want it to behave, the settings will work with syslog-ng, too (but to make sure, do a test run with syslog-ng, also.

I feel, that the syslog-ng side investigation is done now. What do you think? Do you need further assistance?

Cheers,
Attila

**Comment by alltilla** [comment]
Hi @L4rS6 

If the problem occurs with the `ftruncate-test.c` code, too, then it must be an issue with the OS or the file system, most probably the latter.
Unfortunately, I am not equipped to give support for configuring or investigating problems in ZFS. Please reach out to their support channel.

Cheers,
Attila

**Comment by L4rS6** [comment]
Hi @alltilla 

Sure, no problem. Thank you a lot for all the investigation and I'm looking forward for #3689.

Cheers,
Lars

**Comment by alltilla** [comment]
Hi @L4rS6,

Thank you, too, for your extensive support!

If you agree, I would close this issue now.

Cheers,
Attila

**Comment by L4rS6** [comment]
Hi @alltilla, yes I'll close it. Cheers

**Comment by alltilla** [comment]
Hi @L4rS6 ,

It seems like that we might add an option to completely disable truncating, which would fix your problem. It is worth to keep an eye on #3689, but I will keep you updated here. :)

Cheers,
Attila

**Comment by L4rS6** [comment]
Hi @alltilla,

great news, thank you!  I'm already subscribed to the other issue to stay up to date :)

Cheers,
Lars
